cocos/render-scene/core/program-utils.ts

```diff
@@ -114,8 +115,28 @@ export function getVariantKey (programInfo: IProgramInfo, defines: MacroRecord)
     return `${key.toString(16)}|${programInfo.hash}`;
 }
 
+const defaultUniformCounts = new Map<string, number>();
+defaultUniformCounts.set('cc_joints', UBOSkinning.LAYOUT.members[0].count);
+defaultUniformCounts.set('cc_lightPos', UBOForwardLight.LIGHTS_PER_PASS);
+defaultUniformCounts.set('cc_lightColor', UBOForwardLight.LIGHTS_PER_PASS);
+defaultUniformCounts.set('cc_lightSizeRangeAngle', UBOForwardLight.LIGHTS_PER_PASS);
+defaultUniformCounts.set('cc_lightDir', UBOForwardLight.LIGHTS_PER_PASS);
+
+function getUniformSize (prevSize: number, m: Uniform) {
+    if (m.count) {
+        return prevSize + GetTypeSize(m.type) * m.count;
+    } else {
+        const count = defaultUniformCounts.get(m.name);
+        if (count !== undefined) {
+            return prevSize + GetTypeSize(m.type) * count;
+        }
+        console.error(`uniform '${m.name}' must have a count`);
+    }
+    return prevSize;
+}
+
 export function getSize (blockMembers: Uniform[]) {
-    return blockMembers.reduce((s, m) => s + GetTypeSize(m.type) * m.count, 0);
+    return blockMembers.reduce(getUniformSize, 0);
```

Comment:
UniformSize might change according to hardware. We need to get size info at runtime.